EX_browser_sources.c

```diff
@@ -197,66 +204,135 @@ static void Precache_Source(source_data *s)
+static struct curl_buf *curl_buf_init(void)
+{
+	// Q_malloc handles errors and exits on failure
+	struct curl_buf *curl_buf = Q_malloc(sizeof(struct curl_buf));
+	curl_buf->len = 0;
+	curl_buf->ptr = Q_malloc(curl_buf->len + 1);
+	return curl_buf;
+}
+
+static void curl_buf_deinit(struct curl_buf *curl_buf)
+{
+	Q_free(curl_buf->ptr);
+	Q_free(curl_buf);
+}
+
+static size_t curl_write_func( void *ptr, size_t size, size_t nmemb, struct curl_buf *buf )
+{
+	size_t new_len = buf->len + size * nmemb;
+
+	// not checking for realloc errors since Q_realloc will exit on failure
+	buf->ptr = Q_realloc(buf->ptr, new_len + 1);
+
+	memcpy(buf->ptr + buf->len, ptr, size * nmemb);
+	buf->ptr[new_len] = '\0';
+	buf->len = new_len;
+
+	return size * nmemb;
+}
+
+int SB_Cache_Source(const source_data *s, const struct curl_buf *curl_buf)
 {
-	CURL *curl;
-	CURLcode res;
 	size_t filename_buf_len;
 	char *filename;
 	FILE *f;
- 
-	if (s->type != type_url) {
-		Com_Printf_State(PRINT_FAIL, "SB_Update_Source_From_URL() Invalid argument\n");
-		return;
+
+	filename_buf_len = SB_URL_To_Filename_Length(s->address.url);
+	filename = Q_malloc(filename_buf_len);
+	SB_URL_to_FileName(s->address.url, filename, filename_buf_len);
+	if (!FS_FCreateFile(filename, &f, "ezquake/sb/cache", "wb+")) {
+		Com_Printf_State(PRINT_FAIL, "SB_Cache_Source() Can't create cache file");
+		Q_free(filename);
+		return 0;
 	}
 
+	fwrite(curl_buf->ptr, sizeof(char), curl_buf->len, f);
+
+	fclose(f);
+	Q_free(filename);
+	return 1;
+}
+
+struct curl_buf *SB_Retrieve_Data(const source_data *s)
+{
+	CURL *curl;
+	CURLcode res;
+	struct curl_buf *curl_buf;
+
 	curl = curl_easy_init();
 	if (curl) {
 		curl_easy_setopt(curl, CURLOPT_URL, s->address.url);
 		curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
 	}
 	else {
-		Com_Printf_State(PRINT_FAIL, "SB_Update_Source_From_URL() Can't init cURL\n");
-		return;
-	}
-	
-	filename_buf_len = SB_URL_To_Filename_Length(s->address.url);
-	filename = Q_malloc(filename_buf_len);
-	SB_URL_to_FileName(s->address.url, filename, filename_buf_len);
-	if (!FS_FCreateFile(filename, &f, "ezquake/sb/cache", "wt+")) {
-		Com_Printf_State(PRINT_FAIL, "SB_Update_Source_From_URL() Can't open cached file");
-		return;
+		Com_Printf_State(PRINT_FAIL, "SB_Retrieve_Data() Can't init cURL\n");
+		return NULL;
 	}
 
-	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, NULL);
-	curl_easy_setopt(curl, CURLOPT_WRITEDATA, f);
+	curl_buf = curl_buf_init();
+	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, curl_write_func);
+	curl_easy_setopt(curl, CURLOPT_WRITEDATA, curl_buf);
 
 	res = curl_easy_perform(curl);
 	if (res != CURLE_OK) {
-		Com_Printf("Error: Could not read URL %s\n", s->address.url);
+		Com_Printf("SB_Retrieve_Data(): Could not read URL %s\n", s->address.url);
+		curl_easy_cleanup(curl);
+		curl_buf_deinit(curl_buf);
+		return NULL;
 	}
 
-	fseek(f, 0, SEEK_SET);
-	
-	SB_Process_URL_Buffer(f, servers, serversn);
-	fclose(f);
-	Q_free(filename);
+	return curl_buf;
```

Comment:
Add `curl_easy_cleanup(curl);` before returning?

Reply:
Thanks @meag, it's fixed now.